text/tls-and-alpn.rst

```diff
@@ -0,0 +1,380 @@
+
+SSLv2 and SSLv3 are not supported. TLSv1.3 support is included (question:
+shall/can we deliver with OpenSSL >= 1.1.1 consistently for all platforms?).
+
+``--tls-certfile``, ``--tls-keyfile`` and ``--tls-private-key-password`` are
+directly the parameters of Python ``ssl.SSLContext.load_cert_chain()``,
+except that the password cannot be a function. Instead, the EdgeDB server would
+accept password as an environment variable ``EDGEDB_TLS_PRIVATE_KEY_PASSWORD``.
+
+If specified, ``--tls-ciphers`` will be used to call
+``ssl.SSLContext.set_ciphers()``.
+
+We'll discuss later in Core EdgeDB Development for the case of omitting ``tls``
+parameters.
+
+
+CLI Tooling
+-----------
+
+For the ``edgedb`` CLI, the goal is to make TLS in EdgeDB easy to use. There
+are 2 major scenarios:
+
+1. Using EdgeDB with TLS in development process.
+2. Deploy EdgeDB with TLS for production environment.
+
+As we'd like to enable TLS by default, a certificate is always required before
+starting an EdgeDB instance. In brief, for development, a self-signed
+certificate is usually sufficient, so the CLI tooling should have an option to
+automatically generate that for the user transparently. For production, we'd
+also like to make the certificate/key management easier and safe, options are
+discussed later.
+
+
+Internals
+^^^^^^^^^
+
+Under the neath, we'll have new files and folders in the EdgeDB well-known
+folder::
+
+    $WELLKNOWN/
+        +--config/
+        |    +--instance_ports.json
+        |    +--tls.json              <NEW>
+        |--credentials/
+        |    +--<instance>.json       <UPDATED>
+        |--certs/                     <NEW>
+        |    +--<instance>.cert.pem   <NEW>
+        |    +--<instance>.key.pem    <NEW>
+        ...
+
+The ``config/tls.json`` contains optional TLS settings for each instance::
+
+    {
+        "<instance>": {
+            "protocol_min": "<min TLS version>",
+            "protocol_max": "<max TLS version>",
+            "ciphers": "<available ciphers>"
+        },
+        ...
+    }
+
+If an instance has no record in ``config/tls.json``, the EdgeDB server options
+``--tls-protocol-min``, ``--tls-protocol-max`` and ``--tls-ciphers`` will be
+omitted and the default value will be used.
+
+``credentials/<instance>.json`` is updated to include the passphrase for the
+private key file, if any. For example::
+
+    {
+        "port": 10732,
+        "user": "edgedb",
+        "password": "....",
+        "database": "edgedb",
+        "tls_private_key_password": "...."
+    }
+
+Question: I don't feel comfortable storing the key passphrase here in clear
+text, especially if this is in production environment and we asked the user
+for passphrase but stored in this file. Perhaps we could use this JSON value
+only for the development process, and introduce a more secure (thus less
+convenient) way to obtain the private key password, like a customizable command
+as in PostgreSQL?
```

Comment:
Yeah, -1 to storing passwords in clear text for any purposes. We can add an env var for that which is a bit better,

Reply:
1. We don't need this password if we don't have client certificates. Which @fantix said we don't want in another thread.
2. But we need database password. Actually credentials file was created specifically to store secrets and must be saved on the secrets partition, so it doesn't make much sense to move password out of this file. On the contrary, everything other could be moved out from this file to a `EDGEDB_DSN` or dsn specified by other means.